keps/sig-node/3960-pod-lifecycle-sleep-action/README.md

```diff
@@ -0,0 +1,411 @@
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Production Readiness Review Questionnaire](#production-readiness-review-questionnaire)
+  - [Feature Enablement and Rollback](#feature-enablement-and-rollback)
+  - [Rollout, Upgrade and Rollback Planning](#rollout-upgrade-and-rollback-planning)
+  - [Monitoring Requirements](#monitoring-requirements)
+  - [Dependencies](#dependencies)
+  - [Scalability](#scalability)
+  - [Troubleshooting](#troubleshooting)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [x] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [x] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes the addition of a new sleep action for the PreStop lifecycle hook in Kubernetes, allowing containers to pause for a specified duration before termination. This enhancement aims to provide a more straightforward way to manage graceful shutdowns and improve the overall lifecycle management of containers.
+
+An example:
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: lifecycle
+spec:
+  containers:
+  - name: lifecycle-container
+    image: nginx
+    lifecycle:
```

Comment:
I added a explaination below(in summary). We need a deployment&service with this feature to achieve zero downtime in nginx. 
Basicllly: during the rolling update, the kubelet and the Endpoint Controller will handle the termination processes asynchronously, so a sleep prestop can delay the shutdown of the pod and let the endpoints controller remove the terminating pod first.

Reply:
But to do it fully properly you need to drain the connections (send `connection: close`/GOAWAY, etc) which needs app awareness

Reply:
Yes, in order to do it fully properly, we need to exec something like `nginx -s quit`, but I think that's out of the scope of this kep?This feature does not ensure finishing the endpoint termination or drain the connections,  it is only a practical and realistic solution.

Reply:
This kep doesn't aim to specifically handle the nginx situation, it only aims to replace the behaviour that many people use exec to run sleep in their container. You can refer [#114465](https://github.com/kubernetes/kubernetes/issues/114465)

Reply:
I think the issue was created before 1.26 was released that introduced a new default behavior - endpoint slices will stay in `serving` state: https://kubernetes.io/docs/tutorials/services/pods-and-endpoint-termination-flow/ and will not be removed. Before 1.26 just sleeping will ensure the connections have drained naturally. After 1.26 it can go both ways, depending on whether LB is lookng for `ready` or `serving` condition. I'd assume that `serving` will be more and more dominant thus simply introducing a sleep wouldn't work and we need to have sleep AND some custom signalling to the container. Perhaps allowing both - exec hook and sleep hook. Or have an ability to define a sub-hook of a sleep hook.

Reply:
My question is whether sleep alone is ever good without any additional signal sent to the container. I think in many cases sleep paired with the readiness probe start returning false is the best combination.

Is any of 5 happening before the Pod received the `SIGKILL`? My understanding is that after 1.26, endpoint slices will stay around past the Pod deletion in `terminating` but `serving` state. So if traffic is still sent to `terminating` endpoints, sleep is not doing any good. If traffic is not sent to `terminating` endpoint (which was the only possibility before 1.26), than sleep works.

> Sergey: "Ready" means serving+not-terminating. The only time we consider serving+terminating endpoints is when there are ZERO ready endpoints. I don't think that's a factor here.

If this is the behavior expected from many LBs, than `sleep` as suggested is helpful. It would be even more helpful if there is a way to combine `sleep` and a custom signal to the container to move to not-`ready`.

Reply:
> My question is whether sleep alone is ever good without any additional signal sent to the container.

Yes, lots of people use it this way now, successfully.  I am sure that sometimes they lose the race and still serve errors, but it largely works.

> Is any of 5 happening before the Pod received the SIGKILL?

Yes.  Practically all of 1-5 can happen inside of a second.  I added step 6, which is "done".  Moving from 5 to 6 can take tens of seconds.  If you have long, slow operations in your server, you might set termination grace to 5 minutes, but that initial sleep only needs to be 30 seconds, for example.

> if traffic is still sent to terminating endpoints, sleep is not doing any good

Traffic is only sent to terminating endpoints if there are NO OTHER options.  That happens in two cases:
1) the service is going away completely - serving errors is not unexpected
2) the service's deployment in undergoing a rolling update, in which case there will be a better option very soon

> It would be even more helpful if there is a way to combine sleep and a custom signal to the container to move to not-ready

Maybe?  There's really no reason to stop accepting new traffic unless you believe upstream LBs will use that as a signal to pick a different endpoint, which L4 services generally do not in k8s.  It's better to keep accepting.  What's the worst case?  If you accept connections, some of them will work.  If you accept a connection and 1 millisecond later you get SIGKILL, the client gets an error.  If you stop accepting, all connections will get errors.  Some failure is better than all failure.

What we really need is a pod lifecycle state which is explicitly "draining".  This indicates that upstream LBs need to stop sending traffic, and to be more useful than sleep, would require those LBs to positively confirm that they are finished BEFORE sending SIGTERM.  Are you looking for another monumental KEP to shepherd? :)  I'll help...

In the meantime, I think this KEP is an incremental improvement.

Reply:
More I think about usability of `sleep`, more I feel that the small change in design - mark container as NOT READY while sleep is executing, will ensure that there will be no new connections and once sleep is done, there likely will be no active connections left. wdyt?

Reply:
@andrewsykim explained to me that doing this ^^^ will break some other scenarios that terminating endpoints were meant to solve: https://kubernetes.io/blog/2022/12/30/advancements-in-kubernetes-traffic-engineering

So I have an updated proposal - what if we introduce another timeout on the sleep action that indicates that after certain amount of time Container will be marked as not ready.

```
        lifecycle:
          preStop:
            sleep:
              seconds: 30
            overrideReady:
              seconds: 15
```

This way customer app will be even more resilient. It will allow not to terminate requests unexpectedly on shutdown allowing time for the LBs to reconfigure themselves. But also it will also "force" LBs to stop sending traffic at some point and let the connections drain.

I understand that this is not a complete solution. But with this second override I feel more confident that the built-in sleep may be used quite reliably.

Reply:
> mark container as NOT READY while sleep is executing, will ensure that there will be no new connections and once sleep is done, there likely will be no active connections left

I don't buy the "no active connections left" part, but it doesn't matter as long as we don't disrupt them.  I'm not sure I see why making it unready would help?  If anything it would harm reliability of low-N services, becuase (I thinK?) the proxy-terminating-endpoints won't consider unready pods, but it will consider ready ones.

Example:

* t0: Service with `externalTrafficPolicy: Local`.
* t1: LB routes traffic to node, node routes to local pod.
* t2: delete pod
* t3:
  - endpoints controller sees delete and sets `terminating: true`
  - kubelet sees delete and runs preStop- sleep(30)
* t4:
  - kube-proxy sees endpoint update, but will still use it because it's the only option
  - LB controller sees endpoint update and starts programming the LB (slow)
  - pod is marked not-ready by kubelet
* t5:
  - LB routes traffic to node, node routes to local pod (proxy-terminating-endpoints)
  - endpoints controller sees not-ready and sets `serving: false`
* t6: kube-proxy sees endpoint update and can no longer use it (not ready)
* t7: LB routes traffic to node, node has no endpoints = error
* t8: (several seconds later) LB controller stops routing to this node
* t9: sleep concludes
* t10: SIGTERM

Reply:
Woops, should have read it all before replying

> it will also "force" LBs to stop sending traffic at some point

The pod is already terminating - making it unready adds no value that I can see.  One of the mistakes we made early on (and still make with restarts) is to abuse "ready".  "Ready" is totally orthogonal to "terminating".  Because of back-compat, we can't change the meaning of "ready", which is why endpoints has "serving" and "terminating".